go-controller/pkg/ovn/namespace.go

```diff
@@ -466,6 +466,10 @@ func (oc *Controller) createNamespaceAddrSetAllPods(ns string) (addressset.Addre
 					mgmtIfAddr := util.GetNodeManagementIfAddr(hostSubnet)
 					ips = append(ips, mgmtIfAddr.IP)
 				}
+				// Because createNamespaceAddrSetAllPods is called before syncNode, we need to
```

Comment:
I think this is correct (and I actually ended up fixing the same problems on a personal branch of mine), I would however remove this here and have `ensureJoinLRPIPs` call `getJoinLRPAddresses` and lookup the IP in the OVN database in case it doesn't find it when doing `getJoinLRPCacheIPs` (in the cache)

Reply:
To call `getJoinLRPAddresses` inside `ensureJoinLRPIPs` we need the `controller` inside the `joinSwitchIPManager`. I think this does not make the code more readable.
What do you think about this?

Reply:
Not necessarily, you could re-work the code to the following:
```go
// ensureJoinLRPIPs tries to allocate the LRP IPs if it is not yet allocated, then they will be stored in the cache
func (jsIPManager *joinSwitchIPManager) ensureJoinLRPIPs(nodeName string) (gwLRPIPs []*net.IPNet, err error) {
	var ok bool
	// first check the IP cache, return if an entry already exists
	gwLRPIPs, ok = jsIPManager.getJoinLRPCacheIPs(nodeName)
	if ok {
		return gwLRPIPs, nil
	}
	gwLRPIPs = jsIPManager.getJoinLRPAddresses(nodeName)
	if len(gwLRPIPs) > 0 {
		return gwLRPIPs, nil
	}
	gwLRPIPs, err = jsIPManager.lsm.AllocateNextIPs(types.OVNJoinSwitch)
	if err != nil {
		return nil, err
	}

	defer func() {
		if err != nil {
			if relErr := jsIPManager.lsm.ReleaseIPs(types.OVNJoinSwitch, gwLRPIPs); relErr != nil {
				klog.Errorf("Failed to release logical router port IPs %v for node %s: %q",
					util.JoinIPNetIPs(gwLRPIPs, " "), nodeName, relErr)
			}
		}
	}()

	if err = jsIPManager.setJoinLRPCacheIPs(nodeName, gwLRPIPs); err != nil {
		klog.Errorf("Failed to add allocated IPs to the join switch IP cache: %s", err.Error())
		return nil, err
	}

	return gwLRPIPs, nil
}

// getJoinLRPAddresses check if IPs of gateway logical router port are within the join switch IP range, and return them if true.
func (jsIPManager *joinSwitchIPManager) getJoinLRPAddresses(nodeName string) []*net.IPNet {
	// try to get the IPs from the logical router port
	gwLRPIPs := []*net.IPNet{}
	gwLrpName := types.GWRouterToJoinSwitchPrefix + types.GWRouterPrefix + nodeName
	joinSubnets := jsIPManager.lsm.GetSwitchSubnets(nodeName)
	ifAddrs, err := util.GetLRPAddrs(gwLrpName)
	if err == nil {
		for _, ifAddr := range ifAddrs {
			for _, subnet := range joinSubnets {
				if subnet.Contains(ifAddr.IP) {
					gwLRPIPs = append(gwLRPIPs, &net.IPNet{IP: ifAddr.IP, Mask: subnet.Mask})
					break
				}
			}
		}
	}

	if len(gwLRPIPs) != len(joinSubnets) {
		var errStr string
		if len(gwLRPIPs) == 0 {
			errStr = fmt.Sprintf("Failed to get IPs for logical router port %s", gwLrpName)
		} else {
			errStr = fmt.Sprintf("Invalid IPs %s (possibly not in the range of subnet %s)",
				util.JoinIPNetIPs(gwLRPIPs, " "), util.JoinIPNetIPs(joinSubnets, " "))
		}
		klog.Warningf("%s for logical router port %s", errStr, gwLrpName)
		return []*net.IPNet{}
	}
	return gwLRPIPs
}

```

Essentially, make `getJoinLRPAddresses` be a func of `joinSwitchIPManager` instead of `Controller`. There is only one occurrence in the code currently calling `getJoinLRPAddresses` and that can easily be converted from doing `oc.getJoinLRPAddresses` to `oc.joinSwIPManager.getJoinLRPAddresses`. 

This would have `ensureJoinLRPIPs` actually ensure that the items are looked up in a) the cache b) the DB c) allocated if a) and b) retrieve nothing.

Reply:
Thank you very much for your help. I have adjusted the PR. Please review again.

Reply:
Thanks for fixing this!